tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java

```diff
@@ -287,6 +301,17 @@ public synchronized void addInitializer(Op initializer) {
     return addGradients(null, new Output<?>[] {y}, x, null);
   }
 
+  public SaverDef saverDef() {
+    if (saverDef == null) {
+      synchronized (this) {
+        if (saverDef == null) {
+          saverDef = addVariableSaver(this);
```

Comment:
`saverDef` needs to be declared volatile for this to work - http://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html

Reply:
Or we make the whole method synchronized, which also seems fine as it's unlikely to be heavily contended.